azure/Guidelines.md

```diff
@@ -540,6 +546,18 @@ Below is an example of JSON for a Rectangle and Circle:
 ```
 Both Rectangle and Circle have common fields: `kind`, `fillColor`, `lineColor`, and `subscription`. A Rectangle also has `x`, `y`, `width`, and `length` while a Circle has `x`, `y`, and `radius`. The `subscription` is a nested polymorphic type. A `free` subscription has no additional fields and a `paid` subscription has `expiration` and `invoice` fields.
 
+<a name="json-polymorphism-kind-extensible"></a>
+:ballot_box_with_check: **YOU SHOULD** define the kind field of a polymorphic type to be an extensible enum.
+
+<a name="json-polymorphism-kind-immutable"></a>
+:warning: **YOU SHOULD NOT** allow an update (patch) to change the kind field of a polymorphic type.
```

Comment:
Are there actual cases in Azure where this is needed?

Reply:
@johanste would know better than I. 
But, before I joined MS, I worked on a subscription web service customers could change their subscription type from Free to Invoice to Monthly, etc. as there were all polymorphic types. Free had an expiration time period property; Invoice had invoice # and contact email; Subscription had credit card info. So, while I can't recall any specific Azure example, my service would have required this ability.

Reply:
The question is what you do for fields that were previously set on the resource, but that are not valid on the new kind. This is a sometimes a reason to use PUT rather than PATCH.

The scenarios where this has come up is when changing a data source type etc. (where data source types are pluggable)